cpg_workflows/large_cohort/combiner.py

```diff
@@ -1,132 +1,129 @@
+    File listing VDSes in gs:// format, one per line
+
+Raises:
+    ValueError: _description_
+    ValueError: _description_
+    ValueError: _description_
+    ValueError: _description_
+    ValueError: _description_
+    ValueError: _description_
+
+Returns:
+    _type_: _description_
+"""
+
+from itertools import chain
+from typing import TYPE_CHECKING, Any
+
+from graphql import DocumentNode
+
+from hail.vds import VariantDataset, new_combiner, read_vds
 
 from cpg_utils import Path
-from cpg_utils.config import config_retrieve, genome_build, get_config
-from cpg_workflows.inputs import get_multicohort
-from cpg_workflows.targets import SequencingGroup
-from cpg_workflows.utils import can_reuse, exists
+from cpg_utils.hail_batch import init_batch
+from metamist.graphql import gql, query
 
+if TYPE_CHECKING:  # TCH002 https://docs.astral.sh/ruff/rules/typing-only-third-party-import/
+    from hail.vds.combiner.variant_dataset_combiner import VariantDatasetCombiner
 
-def _check_gvcfs(sequencing_groups: list[SequencingGroup]) -> list[SequencingGroup]:
-    """
-    Making sure each sequencing group has a GVCF
+# Get a cohort. Samples already in a VDS will be filtered out from the gvcf combiner list
+# and used as a VDS input
+GET_COHORT_QUERY: DocumentNode = gql(
     """
-    for sequencing_group in sequencing_groups:
-        if not sequencing_group.gvcf:
-            if get_config()['workflow'].get('skip_sgs_with_missing_input', False):
-                logging.warning(f'Skipping {sequencing_group} which is missing GVCF')
-                sequencing_group.active = False
-                continue
-            else:
-                raise ValueError(
-                    f'Sequencing group {sequencing_group} is missing GVCF. '
-                    f'Use workflow/skip_sgs = [] or '
-                    f'workflow/skip_sgs_with_missing_input '
-                    f'to control behaviour',
-                )
-
-        if get_config()['workflow'].get('check_inputs', True):
-            if not exists(sequencing_group.gvcf.path):
-                if get_config()['workflow'].get('skip_sgs_with_missing_input', False):
-                    logging.warning(f'Skipping {sequencing_group} that is missing GVCF {sequencing_group.gvcf.path}')
-                    sequencing_group.active = False
-                else:
-                    raise ValueError(
-                        f'Sequencing group {sequencing_group} is missing GVCF. '
-                        f'Use workflow/skip_sgs = [] or '
-                        f'workflow/skip_sgs_with_missing_input '
-                        f'to control behaviour',
-                    )
-    return [s for s in sequencing_groups if s.active]
-
-
-def check_duplicates(iterable):
+query getCohort($cohort: String!) {
+  cohorts(name: {eq: $cohort}) {
+    sequencingGroups {
+      id
+      analyses(type: {eq: "gvcf"}, status: {eq: COMPLETED}, active: {eq: true}) {
+        output
+      }
+    }
+  }
+}
+""",
+)
```

Comment:
Ah wait... I see. Because this is in a separate runtime from the pipeline workflow, you have to re-query for the Cohort content.

IMO it would be cleaner and easier to do this prep work in the Stage:
- Query for latest VDSs
- Get the SG IDs in those VDSs from the analysis meta (or direct from the object as a fallback)
- Use the already populated Cohort (MultiCohort?) to get the universe of CPG IDs in this analysis
- Back filter all the CPG IDs we have in the VDS out of the to-do list
- Use the already populated SG.gvcf attributes to get the string path to each gVCF that needs adding
- Pass each new gVCF path to include as an argument (or persist to a temp file that's opened inside the combiner job for scalability)
- Pass the VDS path(s) to combine if any, instead of having to re-query from metamist in the Job

This would make the most of the objects already populated when the Stage runs instead of reinventing the wheel by re-querying for the same data.

That would also mean that this Combiner stage has a much tighter scope - "take 0 or more VDS paths, and 0 or more gVCFs, and make a new VDS". That would be much more versatile for code re-use

Reply:
I like this suggestion, I'll do that. Especially since it does clean up the combiner and limit its scope as you say.